packages/grid/x-data-grid/src/hooks/features/filter/useGridFilter.tsx

```diff
@@ -86,13 +112,20 @@ export const useGridFilter = (
         filterModel: filterModel ?? getDefaultGridFilterModel(),
       });
 
-      return {
+      const newState = {
         ...state,
         filter: {
           ...state.filter,
           ...filteringResult,
         },
       };
+
+      const visibleRowsState = getVisibleRowsState(apiRef, newState);
+
+      return {
+        ...newState,
+        visibleRows: visibleRowsState,
+      };
```

Comment:
Nit: Will it be sligthly cleaner to destructure both of them at the same place?
```suggestion
      return {
        ...state,
        filter: {
          ...state.filter,
          ...filteringResult,
        },
        visibleRows: visibleRowsState,
      };
```

Reply:
`newState` is being used to calculate the `visibleRowsState`, so it's not possible to merge everything at once 🙂

Reply:
True, my bad 👍